modules/SMSNotifier/providers/RingCentral.php

```diff
@@ -0,0 +1,157 @@
+     */
+    public $encoding = 'utf-8';
+
+    /**
+     * Format.
+     *
+     * @var string
+     */
+    public $format = 'json';
+
+    /**
+     * Required fields.
+     *
+     * @return string[]
+     */
+    public function getRequiredParams()
+    {
+        return ['CLIENT_ID','RINGUSER','RINGPASS','RINGPHONE','RINGEXT'];
+    }
+
+    public function get($key)
+    {
+        return $this->$key;
+    }
+
+    /**
+     * Response.
+     *
+     * @param Requests_Response $request
+     *
+     * @return bool
+     */
+    public function getResponse(Requests_Response $request)
+    {
+        $response = \App\Json::decode($request->body);
+
+        return isset($response['error']) && !empty($response['error']) ? false : true;
+    }
+
+    /**
+     * Fields to edit in settings.
+     *
+     * @return \Settings_Vtiger_Field_Model[]
+     */
+    public function getSettingsEditFieldsModel()
+    {
+        $fields = [];
+        $moduleName = 'Settings:SMSNotifier';
+        foreach ($this->getRequiredParams() as $name) {
+            $field = ['uitype' => 1, 'column' => $name, 'name' => $name, 'displaytype' => 1, 'typeofdata' => 'V~M', 'presence' => 0, 'isEditableReadOnly' => false];
+            if ($name === 'CLIENT_ID') {
+                $field['text'] = [''];
+                $field['label'] = 'CLIENT_ID';
+                $fields[] = $field;
+                continue;
+            }
+            if ($name === 'RINGUSER') {
+                $field['text'] = [''];
+                $field['label'] = 'RINGUSER';
+                $fields[] = $field;
+                continue;
+            }
+            if ($name === 'RINGPASS') {
+                $field['text'] = [''];
+                $field['label'] = 'RINGPASS';
+                $fields[] = $field;
+                continue;
+            }
+            if ($name === 'RINGPHONE') {
+                $field['text'] = [''];
+                $field['label'] = 'RINGPHONE';
+                $fields[] = $field;
+                continue;
+            }
+            if ($name === 'RINGEXT') {
+                $field['text'] = [''];
+                $field['label'] = 'RINGEXT';
+                $fields[] = $field;
+                continue;
+            }
+        }
+        foreach ($fields as &$field) {
+            $field = Settings_Vtiger_Field_Model::init($moduleName, $field);
+        }
+        return $fields;
+    }
+
+    public function getPatch()
+    {
+        $keys = $this->getRequiredParams();
+        $keys[] = $this->toName;
+        $keys[] = $this->messageName;
+        $params = [];
+        foreach ($keys as $key) {
+            $params[$key] = $this->get($key);
+        }
+
+        return $params;
+
+    }
+
+    public function send() {
+
+        $url = $this->getUrl();
+        $cliend_secret = $this->getAuthorization();
+        $patch = $this->getPatch();
+//
+        $login =  new RingCentral\SDK\SDK($patch['CLIENT_ID'], $cliend_secret, $url);
```

Comment:
What class is this?

Reply:
https://github.com/ringcentral/ringcentral-php

Reply:
We can't merge your commits because an additional library in the system is required, which in most cases won't be used. If you rewrite the communication code to "Guzzle, PHP HTTP client" (here is an example https://github.com/YetiForceCompany/YetiForceCRM/search?q=GuzzleHttp) we will merge your commits.

Reply:
@mariuszkrzaczkowski 

I won't be doing it, as it works just fine in the environment. What is the problem with Ringcentral SDK? It is an official well-maintained library.

you can close the request if you don't see potential in this.
P.S
Letting you know that Ringcentral one of the largest VoIP providers in North America, so potentially including this integration will bring you more clients.

Reply:
@mariuszkrzaczkowski

I am not too sure about guzzle but I do have sms, call logs, and call recordings working in production. Full pbx and mms integration is feasible and I would be willing to create a pull request for complete RingCentral / Telus Business Connect support; though I am not sure what your thoughts are on using curl as its not used in the code base, please let me know. 
```php
// Call Log Auth Example

<?php

$clientId = '';
$clientSecret = '';
$jwt = '';
$url = 'https://platform.ringcentral.com/restapi/oauth/token';
$authorization = base64_encode($clientId . ':' . $clientSecret);
$appurl = 'https://platform.ringcentral.com/restapi/v1.0/account/~/call-log/?view=Detailed';



// Curl OAuth Token

$data = array(
    'grant_type' => 'urn:ietf:params:oauth:grant-type:jwt-bearer',
    'assertion' => $jwt
);


$headers = array(
    'Accept: application/json',
    'Content-Type: application/x-www-form-urlencoded',
    'Authorization: Basic ' . $authorization
);

$curl = curl_init($url);
curl_setopt($curl, CURLOPT_POST, true);
curl_setopt($curl, CURLOPT_POSTFIELDS, http_build_query($data));
curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
curl_setopt($curl, CURLOPT_HTTPHEADER, $headers);

$response = curl_exec($curl);
curl_close($curl);

$resp = json_decode($response);
$token = $resp->access_token;



// Curl Call Logs 

$curl = curl_init($appurl);
curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
curl_setopt($curl, CURLOPT_HTTPHEADER, [
    'Accept: application/json',
    'Authorization: Bearer ' . $token
]);

$response = curl_exec($curl);
curl_close($curl);



// Check If Call Log Already Exists In DB

$data = json_decode($response, true);

if (isset($data['records'])) {
    foreach ($data['records'] as $record) {

    // Query ID in database

        $id = $record['sessionId'];
    ...

```

```php
// Rough MMS / SMS Example

<?php

$from = '';
$to = '';
$text = '';
$attachment = 'data:text/plain;name=test.txt;base64,dGVzdAo=';
$appurl = 'https://platform.devtest.ringcentral.com/restapi/v1.0/account/~/extension/~/mms';

// OAuth Token
...

// Send SMS / MMS

$data = array(
    'from' => array(
        'phoneNumber' => $from
    ),
    'text' => $text,
    'attachments' => array(
        $attachment
    ),
    'to' => array(
        array(
            'phoneNumber' => $to
        )
    )
);

$curl = curl_init($appurl);
curl_setopt($curl, CURLOPT_POST, true);
curl_setopt($curl, CURLOPT_HTTPHEADER, [
    'Accept: application/json',
    'Content-Type: multipart/mixed',
    'Authorization: Bearer ' . $token
]);
curl_setopt($curl, CURLOPT_POSTFIELDS, json_encode($data));
curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);

$response = curl_exec($curl);
curl_close($curl);
```

Reply:
guzzle would be better, it includes default proxy configuration and is simple to use, see
https://github.com/YetiForceCompany/YetiForceCRM/blob/f1904a62fc5cd4d5bb3031b19bb2d090bce49ca6/app/RecordCollectors/UaYouControl.php#L134-L147